cupy/_core/dlpack.pyx

```diff
@@ -20,138 +15,213 @@ import cupy
         device.device_type = kDLROCM
-    device.device_id = dev_id
-
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    if array.dtype.kind == 'u':
-        dtype.code = <uint8_t>kDLUInt
-    elif array.dtype.kind == 'i':
-        dtype.code = <uint8_t>kDLInt
-    elif array.dtype.kind == 'f':
-        dtype.code = <uint8_t>kDLFloat
-    elif array.dtype.kind == 'c':
-        dtype.code = <uint8_t>kDLComplex
-    elif array.dtype.kind == 'b':
-        dtype.code = <uint8_t>kDLBool
+
+    return device
+
+
+# The name of this function is following the framework integration guide of
+# TensorComprehensions.
+cpdef object toDlpack(
+    _ndarray_base array, bint use_versioned=True, bint to_cpu=False,
+    bint ensure_copy=False
+) except +:
+    cdef DLManagedTensor* dlm_tensor
+    cdef DLManagedTensorVersioned* dlm_tensor_ver
+    cdef DLTensor* dl_tensor
+    cdef const char *capsule_name
+
+    # Fetch dtype early (as this can raise a BufferError in theory)
+    cdef uint8_t dtype_code = get_dlpack_dtype_code(array.dtype)
+    cdef size_t dtype_itemsize = array.dtype.itemsize
+
+    # Fetch device information since we need it to deal with CPU logic.
+    cdef DLDevice device = get_dlpack_device(array)
+
+    cdef int32_t ndim = array._shape.size()
+
+    if not to_cpu:
+        owner = array
+    elif not ensure_copy and device.device_type == kDLCUDAManaged:
+        # Managed memory is CPU accessible but consumer may expect kDLCPU.
+        owner = array
+        device.device_type = kDLCPU
+        device.device_id = 0
+    else:
+        # We need to create a CPU copy.  Assumes owner.dtype == array.dtype.
+        # TODO: As noted at calling site, this does not honor the "stream".
+        owner = array.get(stream=None, order='A')
+        device.device_type = kDLCPU
+        device.device_id = 0
+
+    cdef void *dlm_tensor_ptr = stdlib.malloc(
+        sizeof(DLManagedTensorVersioned) if use_versioned
+        else sizeof(DLManagedTensor)
+    )
+    if dlm_tensor_ptr == NULL:
+        raise MemoryError()
+
+    # Note: could coalesc this with the previous allocation in principle
+    cdef int64_t* shape_strides = <int64_t*>stdlib.malloc(
+        ndim * sizeof(int64_t) * 2)
+    if shape_strides == NULL:
+        stdlib.free(dlm_tensor_ptr)
+        raise MemoryError()
+
+    # We need a different setup for versioned/unversioned when it comes to
+    # the context/deleter and additional info in the newer versioned one.
+    if use_versioned:
+        dlm_tensor_ver = <DLManagedTensorVersioned*>dlm_tensor_ptr
+
+        # dl_tensor is identically filled for versioned and unversioned:
+        dl_tensor = &dlm_tensor_ver.dl_tensor
+        capsule_name = CAPSULE_NAME_VER
+
+        dlm_tensor_ver.manager_ctx = <void*>owner
+        cpython.Py_INCREF(owner)
+        dlm_tensor_ver.deleter = deleter_ver
+
+        # "Versioned" specific initialization:
+        dlm_tensor_ver.version.major = 1
+        dlm_tensor_ver.version.minor = 0
```

Comment:
nit: we can avoid hard-coding the values?
```suggestion
        dlm_tensor_ver.version.major = DLPACK_MAJOR_VERSION
        dlm_tensor_ver.version.minor = DLPACK_MINOR_VERSION
```

Reply:
Maybe a `cpdef` for these?  I don't like fetching these directly from the header, since we should be able to update the header without bumping the version incorrectly accidentally.
OTOH, this repeats in a more than 2 places, so probably good to centralize.

Reply:
These macros are exposed to Cython in the dlpack.pxd file. Isn't it the expect behavior (that if we update to a 1.1 header the version exported would be 1.1)? What could be the alternative?

Reply:
If there is a new flag in version 1.1, and we update the header but don't change anything else, we would be wrongly advertising 1.1 when we are still exporting 1.0

Reply:
Thinking about it more, shouldn't this be matching what's requested (i.e. `max_version` passed to `__dlpack__`)? Perhaps having a bool `use_versioned` argument is a bit restrictive?

Alternatively, how about just locking the major version?
```suggestion
        dlm_tensor_ver.version.major = DLPACK_MAJOR_VERSION
        dlm_tensor_ver.version.minor = 0
```

Not sure what `cpdef` solution you had in mind, maybe it follows a similar line?

Reply:
Sorry, will push in a second (already did the echange).  I mean creating a global in this file (rather than getting it from the header).

> Shouldn't this be matching what's requested

No, it must match exactly with what is exported, and that is currently `1, 0`.  The only thing is that we should make sure it is smaller than `max_version` (which it clearly is).

Reply:
Think this was resolved with the addition of

* `IMPL_VER_MAJOR`
* `IMPL_VER_MINOR`

Should we resolve this thread?

If there are still things to discuss, presumably it would be on those `IMPL_VER_*` values

Reply:
Yeah unfortunately I can't resolve the threads (that I created!) myself, and it's been making re-reviews a bit challenging... https://github.com/cupy/cupy/pull/8683#issuecomment-2440480623